### ISSUE TITLE
Relationship between branches

### ISSUE THREAD

**Issue by joshaber** [desktop/desktop#2639]
* Show how the current branch relates to the target branch
* The target is the default branch or the upstream's default branch
* Probably let the user switch the target branch

Designs TBD

**Comment by joshaber** [comment]
(This should probably be split up more once we have designs.)

**Comment by WaleedAshraf** [comment]
Any time estimates for this feature ?

**Comment by joshaber** [comment]
@WaleedAshraf Nope, we're still in the early phases. A lot of design still has to happen before we implement it.

**Comment by donokuda** [comment]
Design time! 😄 Here are a couple problems we are aiming to solve with comparing between two branches:

-  There currently isn't a way to see how far we've diverged from a branch. As a result, it's easy for our working branch to diverge from a default branch (such as master) over time.
- There's no easy to see what kind of changes are about to be brought in. This also adds to the difficulty of not being able to see if there are any overlapping work.

![image](https://user-images.githubusercontent.com/1174461/30655899-24154946-9de7-11e7-90cd-cc1b69334dd3.png)

I started off with a version where I played around with a graph/timeline/subway map visualization to show which commits existed on which branch. The goal at the time was to be able to visualize the result of a merge (as well as understand which specific commits would be brought in:

<img src="https://user-images.githubusercontent.com/1174461/30396666-3ba54c86-987f-11e7-8f18-a7bb7ea5f1a1.png" width="400" />

After some discussion, we've agreed on some of the limitations of that design:

- The graph doesn't offer any value and (not seen in the crop) doesn't accurately represent events such as merges.
- We're not convinced of the value of showing all of the commits
- We're losing precious horizontal space

Fast-forward a few rounds of back and forth designing, we landed on a flow on what we think comparing (and merging upstream changes) could look like:

<img width="892" alt="screen shot 2017-09-20 at 9 27 45 am" src="https://user-images.githubusercontent.com/1174461/30655692-94f5b16a-9de6-11e7-8b80-59d52aa906bb.png">

**Comparing & Merging Flow**

1. When it's time to see how far a working branch has diverged from another, the developer will click into the "Compare" tab (formerly "History"). When there's initially no branch being compared to, we'll display the working branch's history.
2. Clicking on the "Choose a branch..." will open a little dropdown where the developer can search for a branch to compare.
3. After selecting a branch, the list will filter down to only the commits that the working branch is missing from the target branch. From there, the developer can choose to merge those commits into their working branch, or they can stop comparing by selecting "None" in the branch dropdown.

Another problem we plan on tackling for 1.1 is merge conflict. We thought that the merge section could be an opportunity to surface a conflict ahead of time instead of presenting an error after clicking "merge." We will still allow them to attempt a merge, and we'll then kick off into the merge conflict resolution flow.

<img width="658" alt="screen shot 2017-09-20 at 9 28 20 am" src="https://user-images.githubusercontent.com/1174461/30655691-94ee8282-9de6-11e7-8355-d6522ddc2e0d.png">

**Comment by j-f1** [comment]
Looks awesome @donokuda!
1. For showing commits on the current branch that aren’t on the branch being compared with, you could add a second level of tabs with one saying “Ahead :eight:”and the other saying “Behind :four:”
1. It might be useful to show “⬆️ 4 ⬇️ 8” for each branch in the branch picker to help choose which one to compare with.
1. The compare dropdown should have a ✖️ in it when you’ve selected a branch to compare with, allowing you to go back to the history view.
1. “Choose a branch…” should be a combobox, allowing people to type in a custom ref (ex. `master@{1day}`)

**Comment by ikriz** [comment]
It would be nice to see the relationship between more than just two branches.. its often useful to see which branch is the most recent branch without having to compare each branch. In a decent graph this is apparent in the blink of an eye. As j-f1 mentions (point nr 2) visualizing the relationship for all branches would help.

**Comment by mreisz7** [comment]
It might be nice to just have a compressed network view that showed the branch that you're currently on, the branch it branched off of, and then the branch you're targeting if you've opened a pull request targeting a different branch.

**Comment by joshaber** [comment]
This looks like a great start @donokuda!

I _think_ all those mocks address the case where my comparison branch is ahead of my branch. I'm curious if you've thought about the flip side where my branch is ahead.

For me, this is currently the only time I use our History view, and it can be pretty painful. My process is usually something like:

1. "Huh, what did this look like before I changed it?"
1. Scroll through history, hoping I haven't merged `master` in so I can find the commits on my branch.
1. I fastidiously keep my branch up to date, so I can't find my commit.
1. Open the compare page on GitHub.com.

I imagine we'll keep iterating on the "Compare" view so this isn't necessarily something we need to address immediately, but it _is_ a common use case for me.

**Comment by cpcallen** [comment]
IMHO: the former subway-map style timeline display was great, easy to understand, and made it very easy to click through a bunch of different commits on either branch to see what has happened.

I think horizontal is easier to understand than vertical, especially because (in order to be consistent with the list of commits) a vertical timeline needs to run backwards in as you read down the diagram.  That is counterintuitive and confusing.

**Comment by popwebz** [comment]
The subway-map style timeline display is the best way. No need to reinvent the wheel there. There is no sense trying to implement it somewhere else under a menu or tab to save screen real estate. Please add that functionality back. It makes a lot of sense from a UX perspective.

**Comment by kleydon** [comment]
Really missing the comparison graph.

**Comment by jmanwiller** [comment]
Really need that comparison graph - where is the option to compare branches and then merge? Wy was this removed in the first place?

**Comment by nerdneha** [comment]
I'm closing this in favor of our new feature issue on relationship between branches feature work 👉 #3956
